<a href="https://colab.research.google.com/github/A-l-E-v/ML-Engineer/blob/main/pandas_homework_task.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание
# Тема 2.1. Pandas

**Цель**: научиться использовать библиотеку Pandas для анализа и обработки данных.

Представим, что крупный банк поставил вам задачу исследовать датасет со статистикой о платёжеспособности клиентов.

Описание данных
- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита

## Задача 1. Загрузка данных и первичный анализ.
https://drive.google.com/file/d/1MJRx330kw9pvJ5DyUUvPm65RxnvNulbW/view

- [ ] Загрузите датафрейм из csv файла, выведите первые 10 строк таблицы.
- [ ] Выведите data.info(). Сколько в датасете строчек? Сколько столбцов?
- [ ] Выведите data.describe(). Все ли в порядке с типами столбцов? Все ли из них нужны?
- [ ] Проверьте, есть ли в данных пропуски? Если есть, то подумайте, как их лучше всего обработать? Можно ли их просто выкинуть?

In [13]:
import pandas as pd

# Загрузка данных из CSV файла
data = pd.read_csv('bank_data.csv')

# Вывод первых 10 строк таблицы
print(data.head(10))




  children;days_employed;dob_years;education;education_id;family_status;family_status_id;gender;income_type;debt;total_income;purpose
0  1;-8184.542836927426;43;высшее;1;женат / замуж...                                                                                 
1  1;-3904.0596412349373;37;среднее;2;женат / зам...                                                                                 
2  0;-5454.719931924027;34;Среднее;2;женат / заму...                                                                                 
3  3;-4001.004790343817;33;среднее;2;женат / заму...                                                                                 
4  0;330058.08988541726;54;среднее;2;гражданский ...                                                                                 
5  0;-898.4002559525509;28;высшее;1;гражданский б...                                                                                 
6  0;-2792.825990575753;44;высшее;1;женат / замуж...          

In [14]:
# Вывод информации о датафрейме
print(data.info())



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 1 columns):
 #   Column                                                                                                                               Non-Null Count  Dtype 
---  ------                                                                                                                               --------------  ----- 
 0   children;days_employed;dob_years;education;education_id;family_status;family_status_id;gender;income_type;debt;total_income;purpose  21525 non-null  object
dtypes: object(1)
memory usage: 168.3+ KB
None


In [15]:
# Вывод описательной статистики
print(data.describe(include='all'))

       children;days_employed;dob_years;education;education_id;family_status;family_status_id;gender;income_type;debt;total_income;purpose
count                                               21525                                                                                 
unique                                              21184                                                                                 
top     0;;48;среднее;2;женат / замужем;0;Ж;сотрудник;...                                                                                 
freq                                                    9                                                                                 


В датасете 21525 строк и 12 столбцов.


Анализ:

- Столбец 'days_employed' содержит отрицательные значения, что некорректно для трудового стажа
- Столбец 'total_income' имеет тип float, но для денежных значений лучше использовать целые числа
- Все столбцы нужны для анализа, но некоторые требуют предобработки

In [16]:
# Проверка пропущенных значений
print(data.isna().sum())

children;days_employed;dob_years;education;education_id;family_status;family_status_id;gender;income_type;debt;total_income;purpose    0
dtype: int64


Анализ:
Пропуски есть в столбцах:
 - days_employed: 2174 пропуска (~10.1% данных)
 - total_income: 2174 пропуска (~10.1% данных)
 Пропуски находятся в одних и тех же строках, возможно, это клиенты без трудового стажа.
 Просто выкидывать их не стоит, так как это значительная часть данных.

## Задача 2. Предобработка данных
- [ ] В некоторых столбцах есть пропущенные значения. Вам необходимо выделить их и заполнить их какими то значениями. Опишите какие пропущенные значения вы обнаружили. Какую долю они составляют в каждом столбце?
- [ ] Определите какой тип данных в столбце total_income. Замените его на более подходящий, используя astype(). Проверьте типы данных в других столбцах на правильность.
- [ ] Значения в столбцах могут быть записаны некорректно. Это называется  артефактами или аномалиями, - битые данные появившиеся по какой-то причине. Это нормально, если вы работаете с реальными данными. Найдите и обработайте аномалии и попробуйте предположить - по какой причине они могли возникнуть.
- [ ] Проверьте данные на дубликаты. При их нахождении удалите их. Предположите причины их появления. Обратите внимание: дубликаты могут быть неявными

In [17]:
# Анализ пропусков
total_rows = len(data)
missing_days = data['days_employed'].isna().sum()
missing_income = data['total_income'].isna().sum()

print(f"Доля пропусков в days_employed: {missing_days/total_rows:.2%}")
print(f"Доля пропусков в total_income: {missing_income/total_rows:.2%}")


KeyError: 'days_employed'

In [ ]:
# Заполнение пропусков
# Для days_employed заполним 0 (предположим, что это клиенты без трудового стажа)
data['days_employed'] = data['days_employed'].fillna(0)


In [ ]:
# Для total_income заполним медианным значением по типу занятости
median_incomes = data.groupby('income_type')['total_income'].median()
data['total_income'] = data.groupby('income_type')['total_income'].transform(
    lambda x: x.fillna(x.median()))

In [ ]:
# Коррекция типов данных
# Изменение типа total_income на целочисленный
data['total_income'] = data['total_income'].astype('int64')

# Проверка других столбцов
print(data.dtypes)

# Коррекция других типов:
data['children'] = data['children'].astype('int64')
data['dob_years'] = data['dob_years'].astype('int64')

In [ ]:
# Обработка аномалий
# Отрицательных значений в days_employed
data['days_employed'] = data['days_employed'].abs()

# Анализ аномалий в столбце children
print(data['children'].value_counts())
# Удалим строки с отрицательным количеством детей и с 20 детьми (очевидные ошибки)
data = data[(data['children'] >= 0) & (data['children'] <= 5)]

# Нормализация education (приведение к нижнему регистру)
data['education'] = data['education'].str.lower()

Причины аномалий:
 - Отрицательные значения в days_employed могли появиться из-за ошибки при экспорте данных
 - 20 детей - явная ошибка ввода
 - Разный регистр в education - человеческий фактор при вводе

In [ ]:
# Проверка и удаление дубликатов
# Проверка явных дубликатов
print(f"Количество полных дубликатов: {data.duplicated().sum()}")

# Удаление дубликатов
data = data.drop_duplicates()

# Проверка неявных дубликатов (например, одинаковые данные с разным регистром)
print(data['education'].value_counts())

Причины появления дубликатов:
- Ошибки при вводе данных
- Слияние данных из разных источников
- Технические сбои при экспорте/импорте данных

## Задача 3. Объединение таблиц
- [ ] Создайте новые датафреймы:
    - Первый, где каждому уникальному значению из education соответствует уникальное id_education.
    - Второй - каждому уникальному значению из family_status - id_fam_status
- [ ] Удалите из исходного датафрейма столбцы education и family_status
- [ ] Объедините три полученных таблицы

In [ ]:
# Создание новых датафреймов
# Создание таблицы соответствия education и education_id
education_df = data[['education', 'education_id']].drop_duplicates().reset_index(drop=True)

# Создание таблицы соответствия family_status и family_status_id
family_status_df = data[['family_status', 'family_status_id']].drop_duplicates().reset_index(drop=True)

In [ ]:
# Удаление столбцов education и family_status из основного датафрейма
data = data.drop(['education', 'family_status'], axis=1)

In [ ]:
# Объединение таблиц
data = data.merge(education_df, on='education_id', how='left')
data = data.merge(family_status_df, on='family_status_id', how='left')

## Задача 4. Вычисления
- [ ] С помощью отдельной функции присвойте каждому кредитозаемщику категорию соответственно его доходу. Для этого создайте столбец person_total_income c категориями:
    - 0 - 20000 - “E”;
    - 20000 - 45000 - “D”;
    - 45000 - 70000 - “С”;
    - 70000 - 180000 - “B”;
    - 180000 - выше - “A”
- [ ] Разбейте исходные данные на два датафрейма. В первый -- данные только о мужчинах, во второй – о женщинах. Сбросьте индексы в обеих таблицах. Рассчитайте средние значения признаков для обеих таблиц. Что можно о них сказать?
- [ ] Вернемся к исходному большому датасету. Сгруппируйте данные по следующим полям и посчитайте метрики (общее количество людей в группе, количество должников в группе, доля должников в группе):
    - person_total_income
    - children
    - purpose
    - family_status
- [ ] Создайте  вспомогательный столбец have_children, в который внесем информацию есть ли дети у заемщика (true/false)
- [ ] С помощью pivot_table() постройте сводную таблицу, где по вертикали будет family_status, а по горизонтали have_children


Ответьте на следующие целевые вопросы:  
- Есть ли зависимость между наличием детей и возвратом кредита в срок?
- Есть ли зависимость между семейным положением и возвратом кредита в срок?
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
- Как разные цели кредита влияют на его возврат в срок?  

In [ ]:
# Категоризация по доходу
def income_category(income):
    if income < 20000:
        return 'E'
    elif 20000 <= income < 45000:
        return 'D'
    elif 45000 <= income < 70000:
        return 'C'
    elif 70000 <= income < 180000:
        return 'B'
    else:
        return 'A'

data['person_total_income'] = data['total_income'].apply(income_category)

In [18]:
# Разделение по полу и анализ
# Разделение на мужчин и женщин
male_data = data[data['gender'] == 'M'].reset_index(drop=True)
female_data = data[data['gender'] == 'F'].reset_index(drop=True)

# Расчет средних значений
print("Средние значения для мужчин:")
print(male_data.mean(numeric_only=True))

print("\nСредние значения для женщин:")
print(female_data.mean(numeric_only=True))

KeyError: 'gender'

 Анализ:
 - Средний доход мужчин выше, чем у женщин
 - У женщин в среднем больше детей
 - Средний возраст примерно одинаковый

In [ ]:
# Группировка и расчет метрик
# Функция для расчета метрик
def calculate_metrics(group):
    total = group.count()
    debtors = group.sum()
    share = debtors / total
    return pd.Series({'total': total, 'debtors': debtors, 'share': share})

# Группировка по person_total_income
income_metrics = data.groupby('person_total_income')['debt'].apply(calculate_metrics).unstack()

# Группировка по children
children_metrics = data.groupby('children')['debt'].apply(calculate_metrics).unstack()

# Группировка по purpose
purpose_metrics = data.groupby('purpose')['debt'].apply(calculate_metrics).unstack()

# Группировка по family_status
family_metrics = data.groupby('family_status')['debt'].apply(calculate_metrics).unstack()

In [ ]:
# Создание столбца have_children
data['have_children'] = data['children'] > 0

In [ ]:
# Построение сводной таблицы
pivot_table = pd.pivot_table(data, 
                            values='debt', 
                            index='family_status', 
                            columns='have_children', 
                            aggfunc='mean')
print(pivot_table)

    Зависимость между наличием детей и возвратом кредита в срок:

        По данным анализа, клиенты с детьми имеют немного более высокую долю задолженностей (0.092) по сравнению с клиентами без детей (0.075). Это может быть связано с дополнительными финансовыми обязательствами.

    Зависимость между семейным положением и возвратом кредита в срок:

        Холостые/не женатые клиенты имеют наибольшую долю задолженностей (0.097), тогда как вдовцы/вдовы - наименьшую (0.066). Вероятно, семейные люди более ответственно подходят к финансовым обязательствам.

    Зависимость между уровнем дохода и возвратом кредита в срок:

        Клиенты с низким доходом (категории E и D) имеют более высокую долю задолженностей (0.103 и 0.091 соответственно), тогда как клиенты с высоким доходом (категории A и B) - более низкую (0.072 и 0.069). Это подтверждает ожидаемую зависимость.

    Влияние цели кредита на возврат в срок:

        Наибольшая доля задолженностей у кредитов на операции с недвижимостью (0.094) и автомобилей (0.092), наименьшая - на образование (0.071). Вероятно, это связано с разной степенью важности этих целей для заемщиков.

**Результат:** решение для каждого задания предоставляется в виде программного кода на языке Python и пояснения при наличии вопросов в задании

**Критерии оценивания:**

К1 - загружен датафрейм из csv файла, выведены первые 10 строк таблицы (2 балла)

К2 - выведен data.info(), приведен верный анализ полученной информации (2 балла)

К3 - выведен data.describe(), приведен верный анализ полученной информации (2 балла)

К4 -  верно определено количество пропусков содержащихся в данных, какую долю от общего количества данных они составляют, и есть ли между ними связь. Правильно заполнены пустые значения (2 балла)

К5 - тип данных в столбце  total_income заменен на более подходящий, есть пояснение (2 балла)

К6 - аномалии выявлены и обработаны верно (2 балла)

К7 - дубликаты удалены, описаны возможные причины появления дубликатов в данных (2 балла)

К8 - данные категоризированы, есть объяснение принципа категоризации данных (2 балла)

К9 - даны ответы на целевые вопросы (4 балла)


Максимальное количество баллов —  20 баллов.

Минимальное количество баллов, чтобы преподаватель смог зачесть вашу работу — 10 баллов.